### PR TITLE
Bounce & Complaint handling (on AWS only)

### PR DESCRIPTION
**Pull request by niwinz**

### CODE REVIEW

File: backend/src/app/http/awsns.clj [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+(defmethod ig/init-key ::handler
+  [_ cfg]
+  (fn [request]
+    (let [body  (parse-json (slurp (:body request)))
+
+          body  {"Type" "Notification"
+                 "Message" (json/encode-str body)}
+
+          mtype (get body "Type")]
```
Comment (hirunatan) [inline backend/src/app/http/awsns.clj:40]: Wouldn't "mtype" be always "Notification" here? You have set it directly, three lines above.

File: backend/src/app/http/awsns.clj [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+      (let [profile (db/exec-one! conn (sql/select :profile {:id profile-id}))]
+        (when (some #(= (:email profile) (:email %)) (:recipients report))
+          ;; If the report matches the profile email, this means that
+          ;; the report is for itself, can be caused when a user
+          ;; registers with an invalid email or the user email is
+          ;; permanently rejecting receiving the email. In this case we
+          ;; have no option to mark the user as mutted (and in this case
+          ;; the profile will be also inactive.
+          (db/update! conn :profile
+                      {:is-mutted true}
```
Comment (hirunatan) [inline backend/src/app/http/awsns.clj]: 'Muted' word has a single t (i understand that is from the verb 'to mute')

File: backend/src/app/rpc/mutations/profile.clj [hunk 3]
```diff
@@ -366,16 +385,30 @@
...
+                          {:iss :profile-identity
+                           :profile-id (:id profile)})]
 
       (when (not= email (:email profile))
         (check-profile-existence! conn params))
 
+      (when-not (emails/allow-send-emails? conn profile)
+        (ex/raise :type :validation
+                  :code :profile-is-mutted
+                  :hint "looks like the profile has repoted repeatedly as spam or has permanent bounces."))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/profile.clj]: "has been repoRted"

File: backend/src/app/rpc/mutations/profile.clj [hunk 4]
```diff
@@ -366,16 +385,30 @@
...
+
+      (when (emails/has-bounce-reports? conn email)
+        (ex/raise :type :validation
+                  :code :email-has-permanent-bounces
+                  :hint "looks like the email you invite has been repetedly reported as spam or permanent bounce"))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/profile.clj]: repeatedly

File: backend/src/app/rpc/mutations/profile.clj [hunk 5]
```diff
@@ -397,18 +430,33 @@
...
       (when-let [profile (profile/retrieve-profile-data-by-email conn email)]
         (when-not (:is-active profile)
           (ex/raise :type :validation
                     :code :profile-not-verified
                     :hint "the user need to validate profile before recover password"))
+
+        (when-not (emails/allow-send-emails? conn profile)
+          (ex/raise :type :validation
+                    :code :profile-is-mutted
+                    :hint "looks like the profile has repoted repeatedly as spam or has permanent bounces."))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/profile.clj]: "has been repoRted"

File: backend/src/app/rpc/mutations/profile.clj [hunk 6]
```diff
@@ -397,18 +430,33 @@
...
+
+        (when (emails/has-bounce-reports? conn (:email profile))
+          (ex/raise :type :validation
+                    :code :email-has-permanent-bounces
+                    :hint "looks like the email you invite has been repetedly reported as spam or permanent bounce"))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/profile.clj]: repeatedly

File: backend/src/app/rpc/mutations/teams.clj [hunk 7]
```diff
@@ -297,26 +297,48 @@
...
 
       (when-not (some :is-admin perms)
         (ex/raise :type :validation
                   :code :insufficient-permissions))
 
+      ;; First check if the current profile is allowed to send emails.
+      (when-not (emails/allow-send-emails? conn profile)
+        (ex/raise :type :validation
+                  :code :profile-is-mutted
+                  :hint "looks like the profile has repoted repeatedly as spam or has permanent bounces"))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/teams.clj]: "has been repoRted"

File: backend/src/app/rpc/mutations/teams.clj [hunk 8]
```diff
@@ -297,26 +297,48 @@
...
+
+      (when (and member (not (emails/allow-send-emails? conn member)))
+        (ex/raise :type :validation
+                  :code :member-is-mutted
+                  :hint "looks like the profile has repoted repeatedly as spam or has permanent bounces"))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/teams.clj]: "has been repoRted"

Comment (hirunatan) [reply backend/src/app/rpc/mutations/teams.clj]: perhaps you should include in the message what profile is the banned one

File: backend/src/app/rpc/mutations/teams.clj [hunk 9]
```diff
@@ -297,26 +297,48 @@
...
+
+      ;; Secondly check if the invited member email is part of the
+      ;; global spam/bounce report.
+      (when (emails/has-bounce-reports? conn email)
+        (ex/raise :type :validation
+                  :code :email-has-permanent-bounces
+                  :hint "looks like the email you invite has been repetedly reported as spam or permanent bounce"))
```
Comment (hirunatan) [inline backend/src/app/rpc/mutations/teams.clj]: repeatedly

File: backend/src/app/http/awsns.clj (hunk 1 above)
Comment (niwinz) [reply backend/src/app/http/awsns.clj:40]: oh, the previous body definition is just for debug purposes; I forgot delete it.

**Pull request merged**